Add conic-gradient

yiyix opened the pull request:
<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.
-->

- [ ] At least two implementers are interested (and none opposed):
  - [x] Chrome 
  - [x] Firefox: in meetings, implemented behind flags now.
  - [x] WebKit: https://github.com/whatwg/html/issues/5431#issuecomment-641097348
  
- [ ] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://github.com/web-platform-tests/wpt/pull/24299
   * https://github.com/web-platform-tests/wpt/pull/29002
  
- [ ] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   - [x] Chrome: https://bugs.chromium.org/p/chromium/issues/detail?id=1097034
   - [x] Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1627014
   - [x] Safari: https://bugs.webkit.org/show_bug.cgi?id=225539

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/6661/canvas.html" title="Last updated on May 18, 2021, 3:09 PM UTC (f3d6c6e)">/canvas.html</a>  ( <a href="https://whatpr.org/html/6661/6efd0e6...f3d6c6e/canvas.html" title="Last updated on May 18, 2021, 3:09 PM UTC (f3d6c6e)">diff</a> )
<a href="https://whatpr.org/html/6661/infrastructure.html" title="Last updated on May 18, 2021, 3:09 PM UTC (f3d6c6e)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/6661/6efd0e6...f3d6c6e/infrastructure.html" title="Last updated on May 18, 2021, 3:09 PM UTC (f3d6c6e)">diff</a> )

Kaiido commented on source:
```diff
@@ -60353,6 +60354,7 @@ interface mixin <dfn>CanvasFillStrokeStyles</dfn> {
   attribute (DOMString or CanvasGradient or CanvasPattern) <span data-x="dom-context-2d-fillStyle">fillStyle</span>; // (default black)
   <span>CanvasGradient</span> <span data-x="dom-context-2d-createLinearGradient">createLinearGradient</span>(double x0, double y0, double x1, double y1);
   <span>CanvasGradient</span> <span data-x="dom-context-2d-createRadialGradient">createRadialGradient</span>(double x0, double y0, double r0, double x1, double y1, double r1);
+  <span>CanvasGradient</span> <span data-x="dom-context-2d-createConicGradient">createConicFradient</span>(double startAngle, double x, double y);
```
Typo: "createConicFradient" -> "createConicGradient"

Kaiido commented on source:
```diff
@@ -63448,6 +63456,18 @@ try {
+  <p>The <dfn method for="CanvasFillStrokeStyles"><code
+  data-x="dom-context-2d-createConicGradient">createConicGradient(<var>startAngle</var>, <var>x</var>,
+  <var>y</var>)</code></dfn> method takes three arguments, the first argument, <var>startAngle</var>,
+  represents the angle in degree at which begins the gradient, and the last two arguements,
```
Typo: "last two arguements" -> "last two arguments"

yiyix replied:
thanks.

yiyix replied:
fixed.

Kaiido commented on source:
```diff
@@ -63448,6 +63456,18 @@ try {
+  It follows the same rendering rule as CSS <span>conic-gradient</span>. The <var>position</var> is
+  the center defined by <var>x</var> and <var>y</var> in the coordinate space. This function is
+  equivalent to CSS conic-gradient(from <var>startAngle</var> deg at <var>x</var> px <var>y</var> px, <var>angular-color-stop-list</var>),
+  where <var>angular-color-stop-list</var> is to be applied to the <code>CanvasGradient</code>.
```
Should it be specified that it is to be applied by `addColorStop` method?

yiyix replied:
sure! I think it's more clear that way too!

Kaiido reviewed (commented):
Spotted a few typos while reading.

domenic commented on source:
```diff
@@ -63448,6 +63456,19 @@ try {
+  <var>y</var>)</code></dfn> method takes three arguments, the first argument, <var>startAngle</var>,
+  represents the angle in degree at which the gradient begins, and the last two arguments,
+  (<var>x</var>, <var>y</var>), represent the center of the gradient. The method, when invoked, must
+  return a conic <code>CanvasGradient</code> initialized with the specified center.</p>
```
and angle?

yiyix replied:
added

domenic commented on source:
```diff
@@ -63308,8 +63310,8 @@ try {
 
-  <p>There are two types of gradients, linear gradients and radial gradients, both represented by
-  objects implementing the opaque <code>CanvasGradient</code> interface.</p>
+  <p>There are three types of gradients, linear gradients, radial gradients and conic gradients,
```
```suggestion
  <p>There are three types of gradients, linear gradients, radial gradients, and conic gradients,
```

domenic commented on source:
```diff
@@ -63448,6 +63456,19 @@ try {
+  return a conic <code>CanvasGradient</code> initialized with the specified center.</p>
+
+  It follows the same rendering rule as CSS <span>conic-gradient</span>. The <var>position</var> is
+  the center defined by <var>x</var> and <var>y</var> in the coordinate space. This function is
```
I think this sentence about the center is no longer necessary since you give the explicit CSS expression below.

yiyix replied:
i think I wanted to explain what mapped to position. But I agree, it's redundant.

yiyix commented on source:
```diff
@@ -63448,6 +63456,19 @@ try {
+
+  It follows the same rendering rule as CSS <span>conic-gradient</span>. The <var>position</var> is
+  the center defined by <var>x</var> and <var>y</var> in the coordinate space. This function is
+  equivalent to CSS conic-gradient(from <var>startAngle</var> deg at <var>x</var> px <var>y</var> px, <var>angular-color-stop-list</var>),
```
Originally, i didn't put any space, but it looks weird. I can change back.

yiyix commented on source:
```diff
@@ -63448,6 +63456,19 @@ try {
+  It follows the same rendering rule as CSS <span>conic-gradient</span>. The <var>position</var> is
+  the center defined by <var>x</var> and <var>y</var> in the coordinate space. This function is
+  equivalent to CSS conic-gradient(from <var>startAngle</var> deg at <var>x</var> px <var>y</var> px, <var>angular-color-stop-list</var>),
+  where <var>angular-color-stop-list</var> is to be applied to the <code>CanvasGradient</code> by
```
i think it's enough too. the same concept for `addColorStop` is used in both `createLinearGradient` and `createRadicalGradient`. Updated.

yiyix commented on source:
```diff
@@ -3388,6 +3388,7 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
      <li><dfn data-x-href="https://drafts.csswg.org/css-images/#intrinsic-height">intrinsic height</dfn></li>
      <li><dfn data-x-href="https://drafts.csswg.org/css-images/#intrinsic-width">intrinsic width</dfn></li>
      <li>The <dfn data-x-href="https://drafts.csswg.org/css-images-3/#the-image-orientation">'image-orientation'</dfn> property</li>
+     <li><dfn data-x-href="https://drafts.csswg.org/css-images-4/#funcdef-conic-gradient">conic-gradient</dfn></li>
```
added!

Kaiido commented on source:
```diff
@@ -63448,6 +63456,20 @@ try {
+  <p>The <dfn method for="CanvasFillStrokeStyles"><code
+  data-x="dom-context-2d-createConicGradient">createConicGradient(<var>startAngle</var>, <var>x</var>,
+  <var>y</var>)</code></dfn> method takes three arguments, the first argument, <var>startAngle</var>,
+  represents the angle in degree at which the gradient begins, and the last two arguments,
```
I don't see where it has been defined that `angle` should be in degree.
In https://github.com/whatwg/html/issues/5431#issuecomment-608178051 @nt1m was asking if this should be in radians or degrees, but it seems there hasn't been any response in that thread, and given current Nightly still uses radians in the implementation, I guess they haven't been made aware of that decision to use degrees.

For what it's worth, if it's still time to have this discussion, I feel as a web-author that having the angle in degrees is **very** confusing: If I'm not mistaken, it is the only place in the whole canvas API that we would have an angle expressed in degrees instead of in radians. So if we want to play with that angle and e.g the transformation of the context, or the start/end angles of an ellipse, we'd have to convert between degrees and radians, not to mention the mental gymnastic of remembering where to use what...   
If the goal was to have a simpler mapping from CSS, then note that only the &lt;angle> `0` can have the unit omitted, and that CSS has a `rad` unit.

domenic replied:
+1 for radians. I'm not an expert, but all the JS APIs I know of use (e.g. the `Math` functions) use radians, and canvas should definitely be internally consistent.

The inconistency between axes does seem problematic though. Would it make sense to make the mapping `(startAngle + Math.PI / 2)` radians so that it matches other canvas APIs? We could also even add a counterclockwise boolean like those APIs have.

mysteryDate replied:
I implemented this many months ago and cannot remember why I chose degrees. I think I was trying to match some CSS behaviour.
I also would prefer radians, personally. +1 to rotating everything by PI/2 as well to match `arc` and `ellipse`.

yiyix replied:
I am so glad we have this discussion here now before i change more Canvas API. 

+1 for radians and rotating everything by PI/2. Canvas API should be consistent. Changing spec. I will also change our implementation  to map this.

domenic reviewed (approved):
I pushed some editorial tweaks. It now LGTM!

However I noticed that the tests at https://github.com/web-platform-tests/wpt/pull/24299/files seem rather basic. They all use zero angle, so they would not detect the change we made here to radians or to offsetting by pi/2. And if I am reading them right, they all test only a single color (`0,255,0,255`), even though they use multiple color stops. Can we improve the test coverage here?

yiyix commented:
> I pushed some editorial tweaks. It now LGTM!
> 
> However I noticed that the tests at https://github.com/web-platform-tests/wpt/pull/24299/files seem rather basic. They all use zero angle, so they would not detect the change we made here to radians or to offsetting by pi/2. And if I am reading them right, they all test only a single color (`0,255,0,255`), even though they use multiple color stops. Can we improve the test coverage here?

Due to this change, I updated the tests accordingly -> https://chromium-review.googlesource.com/c/chromium/src/+/2891049
I modified the tests a bit, so it's testing for positive/negative angle rotation and 1 test for no-rotation with several colorstop.

I will also add another test for argument validation. What are the other tests you would suggest? I am happy to address them in a later cl.

domenic commented:
I'm still not seeing any tests that actually test the gradient-ness: every pixel test is testing against the same `0,255,0,255` value. So if I am understanding the tests correctly, an implementation would be able to pass these tests if it only ever used the first color stop in the color stop list.

Are there tests that I am missing which test more than one color? Or am I misunderstanding the test suite?

I do see a test under `third_party/blink/web_tests/fast/canvas/conic-gradient.html` that seems like it'd be good, but it's an internal Blink test, not a web platform test. Maybe it could be converted into a [reftest](https://web-platform-tests.org/writing-tests/reftests.html)?

yiyix commented:
> I'm still not seeing any tests that actually test the gradient-ness: every pixel test is testing against the same `0,255,0,255` value. So if I am understanding the tests correctly, an implementation would be able to pass these tests if it only ever used the first color stop in the color stop list.
> 
> Are there tests that I am missing which test more than one color? Or am I misunderstanding the test suite?
> 
> I do see a test under `third_party/blink/web_tests/fast/canvas/conic-gradient.html` that seems like it'd be good, but it's an internal Blink test, not a web platform test. Maybe it could be converted into a [reftest](https://web-platform-tests.org/writing-tests/reftests.html)?

Thank you for the feedbacks! I added much more tests in https://chromium-review.googlesource.com/c/chromium/src/+/2896865

domenic commented:
Excellent, that's perfect! I'm happy to merge this now. I'll give this until Monday in case @annevk or anyone else wants to have a final look, then I'll press the merge button.

yiyix commented:
Added the new tests pull request in the description. https://github.com/web-platform-tests/wpt/pull/29002

annevk reviewed (commented):
Modulo nits this seems okay. At some point we also need to modernize these method definitions.
